optax/losses/_self_supervised_test.py

```diff
@@ -124,6 +124,362 @@ def test_vmap(self, anchor, positive, negative):
+          z2_val,
+          epsilon=eps,
+      )
+      cos_21 = _regression.cosine_similarity(
+          p2_val,
+          z1_val,
+          epsilon=eps,
+      )
+      loss_12 = -cos_12
+      loss_21 = -cos_21
+      loss = 0.5 * (loss_12 + loss_21)
+      return jnp.mean(loss)
+
+    handmade_result = testing_simsiam_loss(p1, z2, p2, z1)
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p1,
+        z2,
+        p2,
+        z1,
+        symmetric=True,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+  def test_single_direction_matches_handmade(self):
+    p = jnp.array(
+        [[0.1, 0.2, 0.3],
+         [0.3, 0.2, 0.1]],
+        dtype=jnp.float32,
+    )
+    z = jnp.array(
+        [[0.4, 0.0, 0.2],
+         [0.1, 0.5, 0.3]],
+        dtype=jnp.float32,
+    )
+
+    def testing_single_direction_simsiam(p_val, z_val, eps=1e-6):
+      cos = _regression.cosine_similarity(
+          p_val,
+          z_val,
+          epsilon=eps,
+      )
+      loss = -cos
+      return jnp.mean(loss)
+
+    handmade_result = testing_single_direction_simsiam(p, z)
+    result = jax.jit(_self_supervised.simsiam_loss)(
+        p,
+        z,
+        symmetric=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-4)
+
+
+class DinoLossTest(absltest.TestCase):
+
+  def test_single_view_matches_handmade(self):
+    student_logits = jnp.array(
+        [[0.1, 0.2, 0.3],
+         [0.3, 0.1, 0.2]],
+        dtype=jnp.float32,
+    )
+    teacher_logits = jnp.array(
+        [[0.0, 0.0, 0.0],
+         [0.2, 0.2, 0.0]],
+        dtype=jnp.float32,
+    )
+    student_temperature = 0.1
+    teacher_temperature = 0.04
+    teacher_center = jnp.array(
+        [0.1, -0.1, 0.0],
+        dtype=jnp.float32,
+    )
+
+    def testing_dino_loss(
+        student,
+        teacher,
+        student_temperature_val,
+        teacher_temperature_val,
+        teacher_center_val,
+    ):
+      teacher_scaled = (
+          teacher - teacher_center_val
+      ) / teacher_temperature_val
+      student_scaled = student / student_temperature_val
+      teacher_prob = jax.nn.softmax(teacher_scaled, axis=-1)
+      log_student_prob = jax.nn.log_softmax(student_scaled, axis=-1)
+      loss_per_example = -jnp.sum(
+          teacher_prob * log_student_prob,
+          axis=-1,
+      )
+      return jnp.mean(loss_per_example)
+
+    handmade_result = testing_dino_loss(
+        student_logits,
+        teacher_logits,
+        student_temperature,
+        teacher_temperature,
+        teacher_center,
+    )
+    result = jax.jit(_self_supervised.dino_loss)(
+        student_logits,
+        teacher_logits,
+        student_temperature=student_temperature,
+        teacher_temperature=teacher_temperature,
+        teacher_center=teacher_center,
+        two_view=False,
+    )
+    np.testing.assert_allclose(result, handmade_result, atol=1e-6)
+
+  def test_two_view_matches_handmade(self):
```

Comment:
rewrite as a random inputs test parameterized by a random seed please (instead of hard-coding the numbers)